src/main/java/org/spongepowered/api/item/recipe/crafting/ShapedCraftingRecipe.java

```diff
@@ -0,0 +1,229 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.recipe.crafting;
+
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.item.ItemTypes;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+import org.spongepowered.api.item.inventory.Slot;
+import org.spongepowered.api.item.inventory.type.GridInventory;
+import org.spongepowered.api.util.ResettableBuilder;
+import org.spongepowered.api.world.World;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+import javax.annotation.Nullable;
+
+/**
+ * A ShapedCraftingRecipe is a CraftingRecipe that has shape and fits into a grid.
+ */
+public interface ShapedCraftingRecipe extends CraftingRecipe {
+    static Builder builder() {
+        return Sponge.getRegistry().createBuilder(Builder.class);
+    }
+
+    /**
+     * @return An unmodifiable copy of the original aisle
+     */
+    List<String> getAisle();
+
+    /**
+     * @return All ingredient predicates; if all of these are fulfilled, the recipe can be crafted
+     */
+    Map<Character, Predicate<ItemStackSnapshot>> getIngredientPredicates();
+
+    /**
+     * @param symbol The symbol paired with the ingredient predicate in the aisle
+     * @return The paired ingredient predicate
+     */
+    default Optional<Predicate<ItemStackSnapshot>> getIngredientPredicate(char symbol) {
+        return Optional.ofNullable(getIngredientPredicates().get(symbol));
+    }
+
+    /**
+     * @param x The x coordinate counted from the left side
+     * @param y The y coordinate counted from the top
+     * @return The ingredient predicate at this position defined by the aisle
+     */
+    default Optional<Predicate<ItemStackSnapshot>> getIngredientPredicate(int x, int y) {
+        if(x < 0 || x >= getWidth() || y < 0 || y >= getHeight())
+            return Optional.empty();
+
+        char symbol = getAisle().get(y).charAt(x);
+
+        return getIngredientPredicate(symbol);
+    }
+
+    @Override
+    default boolean isValid(GridInventory grid, World world) {
+        int gapWidth = grid.getColumns() - getWidth();
+        int gapHeight = grid.getRows() - getHeight();
+
+        // Shift the aisle along the grid wherever possible
+        for(int offsetX = 0; offsetX <= gapWidth; offsetX++) {
```

Comment:
Do I have my checkstyle plugin set up incorrectly? I am not getting the warnings here.